_posts/2024-09-13-string-view-german-style-strings-part-1.md

```diff
@@ -0,0 +1,245 @@
+
+However, Rust-type coercion rules favored a blanket implementation that _did_ copy data. This implementation is shown in the following code block where the `impl<T: AsRef<[u8]>>` will accept any type that implements `AsRef<[u8]>` and copies the data to create a new buffer. To avoid copying, users need to explicitly call `from_vec`, which consumes the `Vec` and transforms it into a buffer.
+
+Diagnosing this implicit copy was time-consuming as it relied on subtle Rust language semantics. We needed to track every step of the data flow to ensure every copy was necessary. To help other users and prevent future mistakes, we also [removed](https://github.com/apache/arrow-rs/pull/6043) the implicit API from arrow-rs in favor of an explicit API. Using this approach, we found and fixed several [other unintentional copies](https://github.com/apache/arrow-rs/pull/6039) in the code base—hopefully, the change will help other [downstream users](https://github.com/spiraldb/vortex/pull/504) avoid unnecessary copies.
+
+
+# Help the Compiler by Giving it More Information
+
+The Rust compiler’s automatic optimizations mostly work very well for a wide variety of use cases, but sometimes, it needs additional hints to generate the most efficient code. When profiling the performance of `view` construction, we found, counterintuitively, that constructing **long** strings was 10x faster than constructing **short** strings, which made short strings slower on StringViewArray than on StringArray!
+
+As described in the first section, StringViewArray treats long and short strings differently. Short strings (&lt;12 bytes) directly inline to the `view` struct, while long strings only inline the first 4 bytes. The code to construct a `view` looks something like this:
+
+It appears that both branches of the code should be fast: they both involve copying at most 16 bytes of data and some memory shift/store operations. How could the branch for short strings be 10x slower?
+
+Looking at the assembly code using [godbolt](https://godbolt.org/), we (with help from [Ao Li](https://github.com/aoli-al)) found the compiler used CPU **load instructions** to copy the fixed-sized 4 bytes to the `view` for long strings, but it calls a function, [`ptr::copy_non_overlapping`](https://doc.rust-lang.org/std/ptr/fn.copy_nonoverlapping.html), to copy the inlined bytes to the <code>view</code> for short strings. The difference is that long strings have a prefix size (4 bytes) known at compile time, so the compiler directly uses efficient CPU instructions. But, since the size of the short string is unknown to the compiler, it has to call the general-purpose function <code>ptr::copy_non_coverlapping</code>. Making a function call is significant unnecessary overhead compared to a CPU copy instruction.
```

Comment:
```suggestion
Looking at the assembly code using [Compiler Explorer](https://godbolt.org/), we (with help from [Ao Li](https://github.com/aoli-al)) found the compiler used CPU **load instructions** to copy the fixed-sized 4 bytes to the `view` for long strings, but it calls a function, [`ptr::copy_non_overlapping`](https://doc.rust-lang.org/std/ptr/fn.copy_nonoverlapping.html), to copy the inlined bytes to the <code>view</code> for short strings. The difference is that long strings have a prefix size (4 bytes) known at compile time, so the compiler directly uses efficient CPU instructions. But, since the size of the short string is unknown to the compiler, it has to call the general-purpose function <code>ptr::copy_non_coverlapping</code>. Making a function call is significant unnecessary overhead compared to a CPU copy instruction.
```
I think the official name of this assembly tool is 'Compiler Explorer', the domain name is the author's name
(And those blogs are great 🚀 )